File: wysicraft-runtime/src/main/java/com/wysicraft/runtime/model/RowTemplates.java
package com.wysicraft.runtime.model;

import java.util.*;
import com.wysicraft.runtime.model.Models.*;

public final class RowTemplates {
    public static final Set<String> TYPES=Set.of("panel","label","image","texture_region","item","button","progress");
    public static final Set<String> ACTIONS=Set.of("","item_click","item_primary","item_secondary");
    public static Ui layout(Element list) { var ui=new Ui(); ui.elements=list.rowElements; return ui; }
    public static void check(Element list) {
        if(list.rowElements.size()>64)throw new IllegalArgumentException("Row template exceeds 64 elements");
        var ui=layout(list);
        for(var e:list.rowElements) {
            if(!TYPES.contains(e.type) || !e.rowTemplate.isEmpty() || !e.rowElements.isEmpty())throw new IllegalArgumentException("Unsupported or recursive row template control");
            if(!ACTIONS.contains(e.rowAction) || !e.rowAction.isEmpty() && !e.type.equals("button"))throw new IllegalArgumentException("Invalid row action");
            if(e.events.values().stream().anyMatch(v->!v.client.actions.isEmpty() || !v.server.actions.isEmpty() || !v.client.script.isEmpty() || !v.server.script.isEmpty()))throw new IllegalArgumentException("Row events belong on the owning list");
            ContainerTree.ancestors(ui,e);
        }
    }
    public static boolean hasAction(Element list,String event) {
        return list.rowElements.stream().anyMatch(e->e.visible && e.enabled && e.rowAction.equals(event)
            && ContainerTree.ancestors(layout(list),e).stream().allMatch(p->p.visible && p.enabled));
    }
    public static boolean hasAction(Element list,String event,Map<String,String> state) {
        return list.rowElements.stream().anyMatch(e->e.rowAction.equals(event) && e.visible && e.enabled && Expressions.evaluate(e.visibleIf,state) && Expressions.evaluate(e.enabledIf,state) && ContainerTree.ancestors(layout(list),e).stream().allMatch(p->p.visible && p.enabled && Expressions.evaluate(p.visibleIf,state) && Expressions.evaluate(p.enabledIf,state)));
    }
    public static Element bind(Element source,ItemRows.Row row,int index) {
        var e=Models.JSON.fromJson(Models.JSON.toJson(source),Element.class);
        e.text=bind(e.text,row,index);e.item=bind(e.item,row,index);e.value=bind(e.value,row,index);e.tooltip=bind(e.tooltip,row,index);return e;
    }
    private static String bind(String s,ItemRows.Row row,int index) {
        return s.replace("${row.item}",row.item()).replace("${row.name}",row.name()).replace("${row.count}",Integer.toString(row.count())).replace("${row.index}",Integer.toString(index));
    }
}

File: src/Wysicraft.Core/RowTemplates.cs
using Wysicraft.Models;
namespace Wysicraft.Core;

/// <summary>Reusable display layouts. Row buttons route to the owning list's trusted events.</summary>
public static class RowTemplates
{
    public static readonly string[] Types = ["panel", "label", "image", "texture_region", "item", "button", "progress"];
    public static readonly string[] Actions = ["", "item_click", "item_primary", "item_secondary"];
    public static List<Element> Resolve(Project project, Element list) => list.RowTemplate.Length == 0
        ? list.RowElements : project.Screens.FirstOrDefault(s => s.Id == list.RowTemplate)?.Elements
            ?? throw new InvalidDataException("Missing row template screen: " + list.RowTemplate);
    public static void Check(IEnumerable<Element> elements)
    {
        var ui = new UiDefinition { Elements = elements.ToList() };
        if (ui.Elements.Count > 64) throw new InvalidDataException("Row templates support up to 64 elements");
        foreach (var e in ui.Elements) {
            if (!Types.Contains(e.Type) || e.RowTemplate.Length > 0 || e.RowElements.Count > 0)
                throw new InvalidDataException("Row templates support panels, labels, images, items, buttons and progress bars; recursive lists are not supported");
            if (!Actions.Contains(e.RowAction) || e.RowAction.Length > 0 && e.Type != "button")
                throw new InvalidDataException("Row action must be an Item List event on a button");
            if (e.Events.Values.Any(v => new[] { v.Client, v.Server }.Any(h => h.Actions.Count > 0 || h.Script.Length > 0)))
                throw new InvalidDataException("Assign row button actions through Row action and the owning Item List events");
            ContainerTree.Ancestors(ui, e).ToArray();
        }
    }
    public static Element Bind(Element source, ItemRow row, int index)
    {
        var e = Json.Clone(source);
        string BindText(string s) => s.Replace("${row.item}", row.Item).Replace("${row.name}", row.Name)
            .Replace("${row.count}", row.Count.ToString(System.Globalization.CultureInfo.InvariantCulture))
            .Replace("${row.index}", index.ToString(System.Globalization.CultureInfo.InvariantCulture));
        e.Text = BindText(e.Text); e.Item = BindText(e.Item); e.Value = BindText(e.Value); e.Tooltip = BindText(e.Tooltip);
        return e;
    }
}
